### PR TITLE
feat(client): joi like validation of relier query parameters

### PR DESCRIPTION
**Pull request by shane-tomlinson**
@vbudhram, @philbooth - all of us have talked about wanting a "joi-like" schema thingy on the client. I put a simple similar API together today. I really like how explicit it makes validation and importing data from all sources. Nothing gets through that's unexpected.

@philbooth - this builds on your original validate function, plus adds a new module that could use a better name.

### CODE REVIEW

File: app/scripts/models/reliers/sync.js [hunk 1]
```diff
@@ -35,19 +45,7 @@ define(function (require, exports, module) {
       var self = this;
       return Relier.prototype.fetch.call(self)
         .then(function () {
-          self.importSearchParam('context');
-
-          try {
-            self.importBooleanSearchParam('customizeSync');
-          } catch (e) {
-            // ignore it for now.
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/sync.js]: startup error handling was completed long ago, `customizeSync` can now be validated.

File: app/scripts/models/mixins/validate.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    createMissingParameterError: function (paramName, Errors) {
+      return toError('MISSING_PARAMETER', paramName, Errors);
+    }
+  };
+
+  function toError(type, param, Errors) {
+    var err = Errors.toError(type);
+    err.param = param;
+    err.message = err.errorModule.toInterpolatedMessage(err, window.translator);
```
Comment (shane-tomlinson) [inline app/scripts/models/mixins/validate.js]: I keep forgetting to remove this.

File: app/scripts/models/reliers/oauth.js [hunk 3]
```diff
@@ -16,9 +16,49 @@ define(function (require, exports, module) {
...
   var Url = require('lib/url');
   var Validate = require('lib/validate');
+  var Validator = require('vat');
 
   var RELIER_FIELDS_IN_RESUME_TOKEN = ['state', 'verificationRedirect'];
 
+  /*eslint-disable camelcase*/
+  var CLIENT_INFO_SCHEMA = {
+    id: Validator.string().required().use(Validate.isHexValid).as('clientId'),
+    image_uri: Validator.string().optional().allow('').use(Validate.isUrlValid).as('imageUri'),
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: @philbooth, @vbudhram - joi marks items optional by default, think we should do the same to reduce boilerplate?

Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: @philbooth, @vbudhram - `use` doesn't seem like the right verb, but joi uses `validate` already to actually do validation. `validate` might go along with `Validate and Transform` though.

File: app/scripts/models/reliers/oauth.js [hunk 4]
```diff
@@ -16,9 +16,49 @@ define(function (require, exports, module) {
...
+    redirectTo: Validator.string().optional().use(Validate.isUriValid),
+    redirect_uri: Validator.string().optional().use(Validate.isUriValid).as('redirectUri'),
+    scope: Validator.string().required().min(1),
+    service: Validator.string().optional(),
+    state: Validator.string().optional(),
+    verification_redirect: Validator.string().optional().use(Validate.isVerificationRedirectValid).as('verificationRedirect')
+  };
+
+  var VERIFICATION_INFO_SCHEMA = {
+    access_type: Validator.string().optional().use(Validate.isAccessTypeValid).as('accessType'),
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: I'd like to add a way to define generic types that are reusable, something like:

```js
Validator.register('access_type', Validator.string().use(Validate.isAcessTypeValid))

...
var VERIFICATION_INFO_SCHEMA = {
  access_type: Validator.access_type().optional().as('accessType'),
...
```

File: app/scripts/models/reliers/oauth.js (hunk 3 above)
Comment (philbooth) [inline app/scripts/models/reliers/oauth.js]: > joi marks items optional by default, think we should do the same to reduce boilerplate?

Sounds good. You could also make it possible to configure the default on initialisation? That way it could be `optional` for people who want that behaviour and `required` for people who want the opposite.

Comment (philbooth) [inline app/scripts/models/reliers/oauth.js]: > `use` doesn't seem like the right verb...

To try and keep things suitable javascripty, how about `apply`? I know it isn't exactly the same concept as `Function.apply`, but it's kind of similar?

Comment (vbudhram) [inline app/scripts/models/reliers/oauth.js]: I agree with marking things optional by default. Instead of `use`, maybe `verify` or `test`?

Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: @philbooth, @vbudhram - thanks for the feedback for both `optional` and `use`. `test` feels like a good fit.

**Comment by shane-tomlinson** [comment]
@philbooth and @vbudhram - I did another round on this, mind some more feedback? The API is coming together much more cleanly. I've also added the ability to vat to register a named validator:

```js
var urlSchema = Vat().string().test(Validate.isUrlValid);
Vat.register('url', urlSchema);

// later on, where a url needs to be validated:
var schema = {
  ...
  image_uri: Vat.url().allow('').as('imageUri')
  ...
};
```

That could simplify the reliers even further, but does require some up front setup, which I haven't figured where should go yet. Think it's a good idea? Any suggestions on where to initialize those types?

File: app/scripts/models/reliers/oauth.js [hunk 5]
```diff
@@ -242,15 +199,9 @@ define(function (require, exports, module) {
...
-          self.set('serviceName', serviceInfo.name);
-
-          self._validateOAuthValues(serviceInfo);
-
-          // server version always takes precedent over the search parameter
-          self.set('redirectUri', serviceInfo.redirect_uri);
-          self.set('termsUri', serviceInfo.terms_uri);
-          self.set('privacyUri', serviceInfo.privacy_uri);
-          self.set('trusted', serviceInfo.trusted);
+          var result = self.validateUsingSchema(
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: `importSearchParamsUsingSchema` takes the schema as the first parameter, should this be consistent here?

File: app/scripts/models/reliers/relier.js [hunk 6]
```diff
@@ -13,31 +13,62 @@
...
+  var QUERY_PARAMETER_SCHEMA = {
+    campaign: Vat.string(),
+    // `email` will be further validated by views to
+    // show the appropriate error message
+    email: Vat.string().allow(Constants.DISALLOW_CACHED_CREDENTIALS),
+    // FxDesktop declares both `entryPoint` (capital P) and
+    // `entrypoint` (lowcase p). Normalize to `entrypoint`.
+    entryPoint: Vat.string(),
+    entrypoint: Vat.string(),
+    migration: Vat.string().valid(Constants.AMO_MIGRATION).valid(Constants.SYNC11_MIGRATION),
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/relier.js]: Think `valid` should be able to take more than one argument at a time to avoid repeated calls to `valid`?

File: app/scripts/models/reliers/sync.js [hunk 7]
```diff
@@ -13,14 +13,24 @@
...
+  var AuthErrors = require('lib/auth-errors');
   var _ = require('underscore');
   var Relier = require('models/reliers/relier');
   var ServiceNameTranslator = require('lib/service-name');
+  var Vat = require('vat');
+
+  /*eslint-disable camelcase*/
+  var QUERY_PARAMETER_SCHEMA = {
+    // context is not available when verifying.
+    context: Vat.string().min(1),
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/sync.js]: `context` seems like it should be on the base relier, and that we should be strict about which values are allowed.

File: app/tests/lib/helpers.js [hunk 8]
```diff
@@ -127,7 +126,7 @@ define(function (require, exports, module) {
       }
     }
 
-    return searchString + pairs.join('&');
+    return '?' + pairs.join('&');
```
Comment (shane-tomlinson) [inline app/tests/lib/helpers.js]: Revert this change, or issue a separate PR for it.

File: app/scripts/models/reliers/relier.js [hunk 9]
```diff
@@ -13,31 +13,62 @@
...
+    preVerifyToken: Vat.string(),
```
Comment (shane-tomlinson) [inline app/scripts/models/reliers/relier.js]: @rfk - do `preVerifyToken`s have a specified format?

Comment (rfk) [inline app/scripts/models/reliers/relier.js]: Yes, they're JWTs, we validate them like this in auth-server:
```
BASE64_JWT = /^(?:[a-zA-Z0-9-_]+[=]{0,2}\.){2}[a-zA-Z0-9-_]+[=]{0,2}$/
preVerifyToken: isA.string().max(2048).regex(BASE64_JWT).optional()
```

File: app/scripts/models/reliers/relier.js (hunk 6 above)
Comment (philbooth) [inline app/scripts/models/reliers/relier.js]: > Think `valid` should be able to take more than one argument at a time to avoid repeated calls to `valid`?

:+1:

**Comment by philbooth** [comment]
> I did another round on this, mind some more feedback?

Looks good, seems like a pretty big win from where I'm standing.

> I've also added the ability to vat to register a named validator... Think it's a good idea?

Yep, I like.

> Any suggestions on where to initialize those types?

Would `app/scripts/lib/constants.js` be suitable? If not, maybe in a new module?

File: app/scripts/models/mixins/search-param.js [hunk 10]
```diff
@@ -16,12 +16,24 @@ define(function (require, exports, module) {
...
     getSearchParam: function (paramName) {
       return Url.searchParam(paramName, this.window.location.search);
     },
 
     /**
+     * Get values from the URL search parameters.
+     *
+     * @param {array of strings} paramNames - name of the search parameters
+     * to get
+     * @returns {object}
```
Comment (vbudhram) [inline app/scripts/models/mixins/search-param.js]: nit: Capitalize `Object`?

File: app/scripts/models/reliers/relier.js (hunk 6 above)
Comment (vbudhram) [inline app/scripts/models/reliers/relier.js]: I dig that, maybe strings can be tested against regex or array of values.

File: app/scripts/models/reliers/relier.js [hunk 11]
```diff
@@ -75,36 +107,20 @@ define(function (require, exports, module) {
...
-          self.importSearchParam('uid');
-          self.importSearchParam('setting');
-          self.importSearchParam('entrypoint');
-          if (! self.has('entrypoint')) {
-            // FxDesktop declares both `entryPoint` (capital P) and
-            // `entrypoint` (lowcase p). Normalize to `entrypoint`.
-            self.importSearchParam('entryPoint', 'entrypoint');
+          // FxDesktop declares both `entryPoint` (capital P) and
+          // `entrypoint` (lowcase p). Normalize to `entrypoint`.
+          if (self.has('entryPoint') && ! self.has('entrypoint')) {
```
Comment (vbudhram) [inline app/scripts/models/reliers/relier.js]: nit: Edge case both `entryPoint` and `entrypoint` exist do nothing? Should one override the other?

File: app/tests/lib/helpers.js [hunk 12]
```diff
@@ -117,7 +117,6 @@ define(function (require, exports, module) {
   }
 
   function toSearchString(obj) {
```
Comment (vbudhram) [inline app/tests/lib/helpers.js]: Minor nit: Does this need to support building out arrays in query string? Or is that an edge case?

File: app/scripts/models/reliers/relier.js (hunk 11 above)
Comment (shane-tomlinson) [inline app/scripts/models/reliers/relier.js]: `entrypoint` overrides `entryPoint` if both exist. I think the logic is equivalent?

File: app/scripts/models/reliers/sync.js (hunk 7 above)
Comment (shane-tomlinson) [inline app/scripts/models/reliers/sync.js]: No, once iframe support is deprecated in the OAuth flow, context will soon by a sync only feature. Keep it here.

File: app/scripts/models/reliers/oauth.js (hunk 5 above)
Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: Leave as is, we validate data using a schema, it makes sense for `data` to come first.

**Comment by shane-tomlinson** [comment]
Before this grows any larger, I'm going to call it and say this is ready for review. @philbooth and @vbudhram  - r?

File: app/scripts/models/reliers/oauth.js [hunk 13]
```diff
@@ -16,9 +16,49 @@ define(function (require, exports, module) {
   var RelierKeys = require('lib/relier-keys');
   var Url = require('lib/url');
   var Validate = require('lib/validate');
+  var Vat = require('lib/vat');
 
   var RELIER_FIELDS_IN_RESUME_TOKEN = ['state', 'verificationRedirect'];
 
+  /*eslint-disable camelcase*/
+  var CLIENT_INFO_SCHEMA = {
+    id: Vat.hex().required().as('clientId'),
```
Comment (philbooth) [inline app/scripts/models/reliers/oauth.js]: I didn't spot this `as(...)` part last time round. So it performs validation and then returns an object where these keys are mapped accordingly (from `clientId` to `id` in this case)?

File: app/scripts/models/mixins/validate.js [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+  module.exports = {
+    /**
+     * Validate data against a schema
+     *
+     * @param {object} data - data to validate
+     * @param {object} schema - schema that can be passed to the validator
+     * @param {object} Errors - Errors module used to create errors
+     * @returns {object} validation/transformation results
+     */
+    validateUsingSchema: function (data, schema, Errors) {
```
Comment (philbooth) [inline app/scripts/models/mixins/validate.js]: In light of the transformation stuff, I wonder if it's possible to come up with a slightly more descriptive name for this, one that indicates data in !== data out?

I haven't come up with anything myself yet, but I'm having a think...

File: app/scripts/models/mixins/validate.js [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+ */
+
+define(function (require, exports, module) {
+  'use strict';
+
+  var Vat = require('vat');
+
+  module.exports = {
+    /**
+     * Validate data against a schema
```
Comment (philbooth) [inline app/scripts/models/mixins/validate.js]: Should also mention the transformation behaviour in these comments.

File: app/scripts/models/reliers/oauth.js (hunk 13 above)
Comment (philbooth) [inline app/scripts/models/reliers/oauth.js]: This is more a comment about your `vat` library, but it is a little bit ambiguous to me as a reader which way round the transformation flows. Is it from `id => clientId` or `clientId => id`? Is there anything you can do to make that clearer?

**Comment by philbooth** [comment]
Seems fine to me, just a couple of trivial naming questions inline. :+1:

File: app/scripts/models/mixins/validate.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+     * @param {string} paramName
+     * @param {object} Errors - Error module
+     * @returns {Error}
+     */
+    createMissingParameterError: function (paramName, Errors) {
+      return toError('MISSING_PARAMETER', paramName, Errors);
+    }
+  };
+
+  function toError(type, param, Errors) {
```
Comment (vbudhram) [inline app/scripts/models/mixins/validate.js]: Would it make sense to move this to `/lib/errors`?

**Comment by vbudhram** [comment]
@shane-tomlinson Looking good! Just minor notes from me.

File: app/scripts/models/reliers/oauth.js (hunk 13 above)
Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: I could rename as to `renameTo` or just simply `rename`. I'm open to suggestions for names, that's why I keep hitting up you and @vbudhram.

Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: I could rename as to `renameTo` or just simply `rename`. I'm open to suggestions for names, that's why I keep hitting up you and @vbudhram.

Comment (philbooth) [inline app/scripts/models/reliers/oauth.js]: Which way round does the transformation go?

File: app/scripts/models/mixins/validate.js (hunk 14 above)
Comment (philbooth) [inline app/scripts/models/mixins/validate.js]: What if the function was called `transformUsingSchema`?

My thinking is that the transformation is the end result and it throws if the schema is not valid, which could kind of be considered an implementation detail rather than the main purpose. What do you reckon?

File: app/scripts/models/reliers/oauth.js (hunk 13 above)
Comment (shane-tomlinson) [inline app/scripts/models/reliers/oauth.js]: key is source, target is specified in `as` or `renameTo` or `rename`

Comment (vbudhram) [inline app/scripts/models/reliers/oauth.js]: Would `transformTo` make sense, `Vat.hex().required().transformTo('clientId')`. Both `renameTo` and `rename` sound good to me as well.

Comment (philbooth) [inline app/scripts/models/reliers/oauth.js]: I prefer `renameTo` to the simpler `rename` because the direction is the bit that confused me. Anything with the `To` suffix sounds good to me.

File: app/scripts/models/mixins/validate.js (hunk 14 above)
Comment (shane-tomlinson) [inline app/scripts/models/mixins/validate.js]: I like this suggestion. Renaming.

**Comment by philbooth** [comment]
Looks good. :+1:
